src/common/ceph_posix_fallocate.c

```diff
@@ -0,0 +1,65 @@
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <errno.h>
+
+// On FreeBSD ZFS fallocate always fails since it is considered impossible to
+// reserve space on a COW filesystem. It returns EINVAL
+// Linux in this case already emulates it in glibc
+// In which case it is allocated manually, and still that is not a real guarantee
+// that a full buffer is allocated on disk, since it could be compressed.
+// To prevent this the written buffer needs to be loaded with random data.
+
+int manual_fallocate(int fd, off_t offset, off_t len) {
+    // Try to manually allocate the buffer
+    char data[1024*128];
+    int r;
+    r = lseek(fd, offset, SEEK_SET);
+    if (r == -1)
+	return -errno;
+    for (uint64_t off = 0; off < len; off += sizeof(data)) {
+      if (off + sizeof(data) > len)
+        r = write(fd, data, len - off);
+      else
+        r = write(fd, data, sizeof(data));
+      if (r == -1) {
+      	 return -errno;
+      }
+    }
+    return 0;
+}
+
+int ceph_posix_fallocate(int fd, off_t offset, off_t len) {
+
+    int ret = -1;
+
+#ifdef HAVE_POSIX_FALLOCATE
+    ret = posix_fallocate(fd, offset, len);
+    ret = -ret;
+#if !defined(__FREEBSD__)
+    return ret;
+#else
+    if ( ret != !EINVAL ) {
+      return ret;
+    }
+#endif
+#elif defined(__APPLE__)
+    fstore_t store;
+    store.fst_flags = F_ALLOCATECONTIG;
+    store.fst_posmode = F_PEOFPOSMODE;
+    store.fst_offset = offset;
+    store.fst_length = len;
+
+    ret = fcntl(fd, F_PREALLOCATE, &store);
+    if (ret == -1) {
+      ret = -errno;
+    }
+#endif
+    if (ret < 0) {
```

Comment:
this is dead code if HAVE_POSIX_FALLOCATE and not FreeBSD, also, the above __FREEBSD__ and __APPLE__ blocks are difficult to read. please consider restructuring them.

Reply:
Right it is.
However trying to restructure this in the case that we do have `HAVE_POSIX_FALLOCATE` and `__FreeBSD__` running on ZFS things start to get hairy because compile-time and run-time start to interact. But I'll do another attempt.